eggtimer.js

```diff
@@ -1,326 +1,681 @@
 const fs = require('fs');
 const http = require('http');
-const githubWebhookHandler = require('github-webhook-handler');
+const createHandler = require('github-webhook-handler');
 const nodeGithub = require('github');
+const Promise = require("bluebird");
+const assert = require('assert');
+const endOfLine = require('os').EOL;
 
-///////////////////////////////////////////////////////////////////////////////////////////////////
-// Setup
-///////////////////////////////////////////////////////////////////////////////////////////////////
-
-const CONFIG = JSON.parse(fs.readFileSync('config.js'));
-const HANDLER = githubWebhookHandler({ path: CONFIG.github_webhook_path, secret: CONFIG.github_webhook_secret });
-const GITHUB = new nodeGithub({ version: "3.0.0" });
-const GITHUB_AUTHENTICATION = { type: 'token', username: CONFIG.github_username, token: CONFIG.github_token };
-
-///////////////////////////////////////////////////////////////////////////////////////////////////
-// PR state representation
-///////////////////////////////////////////////////////////////////////////////////////////////////
-
-// PRs contains status about incomplete pr's:
-// {
-//     'https://api.github.com/repos/dgmltn/api-test/pulls/5': {
-//         head_sha: 'abcd1234...',
-//         ref: 'my-pull-request',
-//         checks: {
-//             'context1': true|false,
-//             'context2': true|false
-//         },
-//         reviews: {
-//             'user1': true|false,
-//             'user2': true|false
-//         },
-//         mergeable: true|false
-//     }
-// }
-var prs = {};
-
-// commits references a pr url to a commit sha:
-// {
-//     'abcd1234...': 'https://github.com/dgmltn/api-test/pull/5',
-// }
-var commits = {};
-
-///////////////////////////////////////////////////////////////////////////////////////////////////
-// Webhook Handlers
-///////////////////////////////////////////////////////////////////////////////////////////////////
-
-http.createServer(function (req, res) {
-  HANDLER(req, res, function (err) {
-    res.statusCode = 404
-    res.end('no such location')
-  });
-}).listen(CONFIG.port);
+const Config = JSON.parse(fs.readFileSync('config.js'));
+const WebhookHandler = createHandler({ path: Config.github_webhook_path, secret: Config.github_webhook_secret });
+const Github = new nodeGithub({ version: "3.0.0" });
+const GithubAuthentication = { type: 'token', username: Config.github_username, token: Config.github_token };
+
+const MergeFailedLabel = "S-merge-failed";
```

Comment:
A label marking stuck/unmergeable PRs would be useful indeed. In fact, it may be _required_ for the correct bot operation. This is not such a label AFAICT because it marks transient, not permanent failures (i.e., it is ignored/reset by the bot).